rest/diagnostic_api.go

```diff
@@ -0,0 +1,347 @@
+				}
+			}
+		}
+	}
+
+	// Loop over current and past channels
+	for scope, collectionConfig := range user.GetCollectionsAccess() {
+		if _, ok := h.db.CollectionNames[scope]; !ok {
+			continue
+		}
+		for collectionName, CAConfig := range collectionConfig {
+			if _, ok := h.db.CollectionNames[scope][collectionName]; !ok {
+				continue
+			}
+			keyspace := scope + "." + collectionName
+			resp.AdminGrants[keyspace] = make(map[string]auth.GrantHistory)
+			resp.DynamicGrants[keyspace] = make(map[string]auth.GrantHistory)
+
+			for channel, chanHistory := range CAConfig.ChannelHistory() {
+				resp.addGrants(chanHistory.Source, keyspace, channel, chanHistory)
+			}
+			for channel, chanEntry := range CAConfig.Channels() {
+				history := auth.GrantHistory{Entries: []auth.GrantHistorySequencePair{{StartSeq: chanEntry.VbSequence.Sequence}}}
+				// If channel is in history, add grant history to current info
+				if _, ok := CAConfig.ChannelHistory_[channel]; ok {
+					for _, entry := range CAConfig.ChannelHistory_[channel].Entries {
+						history.Entries = append(history.Entries, entry)
+					}
+				}
+				resp.addGrants(chanEntry.Source, keyspace, channel, history)
+			}
+		}
+	}
+
+	// Single named collection + default collection handling, if all_channels includes more than one channel ("!"), its using the default collection.
+	if len(user.ExplicitChannels().AllKeys()) != 0 {
+		err = authenticator.RebuildCollectionChannels(userPrinc, "_default", "_default")
+		if err != nil {
+			return err
+		}
+
+		resp.AdminGrants["_default._default"] = make(map[string]auth.GrantHistory)
+		resp.DynamicGrants["_default._default"] = make(map[string]auth.GrantHistory)
+		for channel, chanHistory := range user.ChannelHistory() {
+			resp.addGrants(chanHistory.Source, "_default._default", channel, chanHistory)
+		}
+		for channel, chanEntry := range user.Channels() {
+			grantInfo := auth.GrantHistory{Entries: []auth.GrantHistorySequencePair{{StartSeq: chanEntry.VbSequence.Sequence}}}
+			if chanHistory, ok := user.ChannelHistory()[channel]; ok {
+				for _, entry := range chanHistory.Entries {
+					grantInfo.Entries = append(grantInfo.Entries, entry)
+				}
+			}
+			resp.addGrants(chanEntry.Source, "_default._default", channel, grantInfo)
+		}
+	}
+
+	bytes, err := base.JSONMarshal(resp)
+	if err != nil {
+		return err
+	}
+	h.writeRawJSON(bytes)
+	return err
+}
+
+func (resp getAllChannelsResponse) addGrants(source string, keyspace string, channelName string, grantInfo auth.GrantHistory) {
+	if source == channels.AdminGrant {
+		resp.AdminGrants[keyspace][channelName] = grantInfo
+	} else if source == channels.DynamicGrant {
+		resp.DynamicGrants[keyspace][channelName] = grantInfo
+	} else if source == channels.JWTGrant {
+		resp.JWTGrants[keyspace][channelName] = grantInfo
+	}
+	grantInfo.Source = ""
+}
+
+func (resp getAllChannelsResponse) addRoleGrants(roleName string, source string, keyspace string, channelName string, grantInfo auth.GrantHistory) {
```

Comment:
As above:

This should be a pointer. This modifies the object and this works even though this is a copy because the objects modified are pointers:

```suggestion
func (resp *getAllChannelsResponse) addRoleGrants(roleName string, source string, keyspace string, channelName string, grantInfo auth.GrantHistory) {
```